### PR TITLE
Port CE geometry preparation from solvcon to `EulerCore`

### LINKED ISSUE #66: Add tests and code for 2D mixed elements
Port 2D mixed-element (triangle + quadrilateral) CE/SE metric computation from https://github.com/solvcon/solvcon .  The work is also associated with issue #404.

The issue tracking 3D mesh is #67.

Tasks:
- [x] Phase 1: Mesh and CE Tests for 2D Mixed Elements
  - PR #859.
  - PR #862.
  - PR #865.
- [x] Phase 2: Gradient Element Geometry
  - Phase 2.1: PR #866.
  - Phase 2.2: PR #869.
- [x] Phase 3: Solution Storage and Initialization
  - PR #870.
- [x] Phase 4: Solution Marching
  - PR #873.
- [x] Phase 5: Boundary Conditions
  - PR #874.

---

### Phase 1: Mesh and CE Tests for 2D Mixed Elements

Add quad-only and mixed tri/quad test coverage for `StaticMesh` and `EulerCore::prepare_ce_2d()`.  Core deliverable for issue #66.

**Files:** `tests/test_mesh_2d.py` (new), `tests/test_euler_multidim.py` (extend), `tests/data/mixed_tq.msh` (new).

**Mesh tests (`test_mesh_2d.py`):**

- Single quad (unit square): face count, normals, areas, `clvol`, `clcnd`, boundary edges, ghost cells.
- 2x2 quad grid (9 nodes, 4 cells): interior/boundary face split, `fccls`, all `clvol` = 0.25.
- Mixed 2 tri + 1 quad (6 nodes): `cltpn` mix, face extraction, `clvol` sum.
- Gmsh import of `mixed_tq.msh`: counts, spot-check geometry.
- Single-cell edge cases; `use_incenter()` with triangles.

**CE tests (`test_euler_multidim.py`):**

- `EulerCoreCEQuadTC`: BCE sum == CCE, hand-verified BCE for unit square, sub-face centroids/normals.
- `EulerCoreCEMixedTC`: varying `clfcs` counts, positive CCE, BCE == CCE.
- `rectangle.msh` regression: total CE volume approx. domain area.

---

#### Phase 2.1: Port `GradientElement` from solvcon.

**Port from:**

- [GradientElement.hpp](https://github.com/solvcon/solvcon/blob/1e0eb87/libmarch/include/march/mesh/ConservationElement/GradientElement.hpp) -- gradient evaluation points, displacement matrices, per-cell tau.
- [BasicCE.hpp](https://github.com/solvcon/solvcon/blob/1e0eb87/libmarch/include/march/mesh/ConservationElement/BasicCE.hpp) -- reference for sub-face geometry used by gradient elements.

**Files (actual):** `GradientElement.{hpp,cpp}` (new), `CMakeLists.txt`, `wrap_multidim.cpp`, `test_multidim.py` (new).

**Results:**
- `GradientElement(mesh, cecnd, icl, tau)` is built on the fly per cell.
- Added `GradientElementType` / `GradientElementTypeGroup` (`getype()`) with sub-element face tables for quad/tri/hex/tet/prism/pyramid; dimension-generic (2D + 3D).

**Tests:** gradient points inside cell, displacement matrix non-singular, `tau = 0` gives CE centroid, individual 2D/3D cell and mixed 2D mesh.

#### Phase 2.2: Populate `GradientElement` Geometry

Give `GradientElement` the displacement-matrix build + solve so a caller can reconstruct a gradient from supplied solution deltas.  (Geometry + linear algebra only.)

**Port from:**

- [GradientElement.hpp](https://github.com/solvcon/solvcon/blob/1e0eb87/libmarch/include/march/mesh/ConservationElement/GradientElement.hpp) -- `calc_displacement_matrix(ifge)`, `nfge`, `nfge_inverse`, `faces`.
- [derivative.hpp](https://github.com/solvcon/solvcon/blob/1e0eb87/libmarch/include/march/gas/derivative.hpp) -- per-FGE solve `grad = adjugate(dst) * udf / det(dst)`.
- [Matrix.hpp](https://github.com/solvcon/solvcon/blob/1e0eb87/libmarch/include/march/core/Matrix.hpp) -- `volume` (determinant), `unnormalized_inverse` (adjugate), and `product` (matrix-vector) for the 2D/3D per-FGE solve.

**Files (actual):** `GradientElement.{hpp,cpp}`, `wrap_multidim.cpp`, `gtests/test_nopython_multidim.cpp` (new, suite `Multidim`), `gtests/CMakeLists.txt`, `test_multidim.py`, `StaticMesh_boundary.cpp` (the `fill_ghost` fix).

**New methods:** `nfge()`, `nfge_inverse()`, `faces(ifge)`, `displacement_matrix(ifge)`, `solve_gradient(ifge, udf)`.

**Results:**
- `solve_gradient(ifge, udf)` reconstructs a per-FGE gradient as `grad = adjugate(dst) * udf / det(dst)`, where `dst = displacement_matrix(ifge)`; `GradientElementType` gains `nfge_inverse`.
- No `small_matrix.hpp`: the 2D/3D determinant, adjugate, and matrix-vector routines are `static` members of `GradientElement` (`determinant`, `adjugate`, `multiply`) over the nested `ge_vector_type` / `ge_matrix_type` aliases, keeping the small dense linear algebra inside its only consumer and avoiding namespace-level names confusable with the dynamic `modmesh::small_vector` container.
- The pybind11 bindings are split into `wrap_management` (construction and read-only scalar properties) and `wrap_gradient` (per-face and per-FGE accessors), mirroring `WrapEulerCore`.
- Fixed the `fill_ghost` 3D ghost-face metric bug (`StaticMesh::fill_ghost` normalized the 3D ghost-face normal by a stray single-index `m_fcnml(ifc)` and stored the area into the normal array, leaving single-cell 3D CE geometry NaN, the root cause of the previously skipped 3D tests); it is committed first, ahead of the gradient-element work.

**Tests:** displacement matrix non-singular per cell type; analytic linear field reconstructed exactly from known deltas (C++ gtest + Python); hand-computed 2D/3D determinant, adjugate, and matrix-vector checks (`Multidim.ge_linalg_2d` / `ge_linalg_3d`).  Both 2D and 3D cells are exercised; the 3D value tests run (no longer skipped) thanks to the `fill_ghost` fix.

---

### Phase 3: Solution Storage and Initialization

Add the conserved-variable / gradient tables and uniform-flow initialization to
`EulerCore`, plus the CFL estimate.  Implemented in draft PR #870 (branch
`feat/solution`); the as-built result is recorded below.

**Port from:**

- [Solution.hpp](https://github.com/solvcon/solvcon/blob/1e0eb87/libmarch/include/march/gas/Solution.hpp) -- `SolutionTableTraits` (`neq = NDIM + 2`), the table set `m_so0c`/`m_so0n`/`m_so0t`/`m_so1c`/`m_so1n`/`m_stm`/`m_cflo`/`m_cflc`/`m_gamma` and their per-row layout, and `update()` (swaps **only** `so0c<->so0n` and `so1c<->so1n`).
- [Solver_decl.hpp](https://github.com/solvcon/solvcon/blob/1e0eb87/libmarch/include/march/gas/Solver_decl.hpp) -- `init_solution`/`calc_cfl` declarations and `TINY = 1.e-60`.
- [Solver.hpp](https://github.com/solvcon/solvcon/blob/1e0eb87/libmarch/include/march/gas/Solver.hpp) -- `init_solution` (`Solver.hpp:227-236`) and `calc_cfl` (`Solver.hpp:147-178`: `hdt = time_increment/2`, `dist` = min CCE->BCE centroid distance, `wspd = sqrt(gamma*p/rho) + |v|`, `cflo = hdt*wspd/dist`, the `cflc` clamp, and the negative-pressure energy correction).

**Files (actual):** `euler_solution.cpp` (new -- `initialize_solution`, `init_solution`, `calc_cfl`), `euler.hpp` (members, getters, `neq()`, inline `update()`), `euler_ce.cpp` (ctor: initialize `m_neq`, call `initialize_solution()`), `cpp/modmesh/multidim/CMakeLists.txt` (`euler_solution.cpp` in `MODMESH_MULTIDIM_SOURCES`), `pymod/wrap_multidim.cpp`, `tests/test_multidim.py`.

**New arrays** (allocated in a new private `initialize_solution()` via an `alloc` lambda -- `arr = SimpleArray<real_type>(shape, 0); arr.set_nghost(m_ngstcell)`; first axis `total = ngstcell + ncell`, so every array carries ghost rows; `neq = ndim + 2` cached in `m_neq`):

- `m_so0c`, `m_so0n`, `m_so0t`, `m_stm` -- shape `{total, neq}`.
- `m_so1c`, `m_so1n` -- shape `{total, neq, ndim}` (true 3D `SimpleArray`, like `m_sfcnd` / `m_sfnml`; solvcon stores the order-1 table flat as `ndim*neq`, so this 3D layout is a deliberate modmesh choice that the Phase 4 gradient indexing must follow).
- `m_cflo`, `m_cflc`, `m_gamma` -- shape `{total}`.

Each has a getter `SimpleArray<real_type> & NAME()`; `neq() const { return m_neq; }`.

**Results:**

- **No `Hand` / `LookupTable` machinery** (confirmed).  Plain `SimpleArray<real_type>` members indexed by column; only the column convention is ported (col `0` = density, cols `1 .. ndim` = momentum, col `m_neq-1` = energy).  `EulerCore` stays one non-templated class with runtime `m_ndim` serving both 2D and 3D; `m_neq` is cached in the ctor (`mesh->ndim() + 2`).
- **`init_solution(real_type gamma, real_type rho, std::array<real_type, 3> const & velocity, real_type p)`** -- uniform primitive-to-conserved fill over real cells `[0, ncell)`: `so0n(icl,0)=rho`, `so0n(icl,1+d)=rho*velocity[d]`, `so0n(icl,m_neq-1)=p/(gamma-1)+0.5*rho*|velocity|^2`; only the leading `ndim` entries of `velocity` are read and the `so0n` ghost rows are left untouched.  `m_gamma.fill(gamma)` fills **every** row (ghost included).  Validates `gamma > 1`, `rho > 0`, `p >= 0` (throws `std::invalid_argument`).  This is the deliberate divergence from solvcon's `Order0Hand::set_by(gas_constant, gamma, density, temperature)`, which zeroes momentum, sets `E = p/(gamma-1)` (quiescent gas), and never writes the gamma table: a moving uniform inflow (issue #404) needs nonzero momentum, kinetic energy, and a populated `m_gamma` (which `calc_cfl` divides by `gamma-1`).
- **`calc_cfl()`** -- per real cell: `hdt = m_time_increment/2`; `dist` = min over faces `ifl` of `|cecnd(icl, ifl*ndim + d) - cecnd(icl, d)|` (CCE-to-BCE centroids, read from `m_cecnd`); `ke = momsq/(2*rho)`, `pr = (ga-1)*(E-ke)`, `pr_adj = (pr+|pr|)/2`, `wspd = sqrt(ga*pr_adj/rho) + sqrt(momsq)/rho`; `cflo(icl) = hdt*wspd/dist`; `cflc(icl) = (cflo-1)*pr_adj/(pr_adj+TINY) + 1`; then rewrites the stored energy to `pr_adj/(ga-1) + ke + TINY`.  `TINY = 1.e-60` is a local `constexpr` in the function.
- **`update()`** -- inline in `euler.hpp`: `m_so0c.swap(m_so0n); m_so1c.swap(m_so1n);` (order-0 and order-1 current/new buffers only; `so0t`, `stm`, `cflo`, `cflc`, `gamma` are not swapped).
- **Bindings** -- `neq` via `wrap_management` (`def_property_readonly`); `init_solution` / `calc_cfl` / `update` via `wrap_preparation` (`.def_timed`); the Python `init_solution` takes a velocity **list** `v`, which the binding length-checks against `ndim` and copies into the `std::array<3>` before calling the C++ method; the nine arrays via `wrap_array` (`MM_DECL_ARRAY`).

**Tests (actual, `tests/test_multidim.py`):** a `_EulerSolutionBase` mixin -- `test_solution_array_shapes` (`neq` and the `(total, neq)` / `(total, neq, ndim)` / `(total,)` shapes), `test_init_solution_columns` (column values + pressure recovery; ghost rows untouched), `test_init_solution_validation` (bad `gamma` / `rho` / `p` raise), `test_calc_cfl_uniform_field` (closed-form `cflo`), `test_calc_cfl_negative_pressure` (energy clamp), `test_update_swaps_buffers` -- run across triangle / quad / mixed / tetrahedron / hexahedron via `EulerSolutionTriangleTC`, `EulerSolutionQuadTC`, `EulerSolutionMixedTC`, `EulerSolutionTetrahedronTC`, `EulerSolutionHexahedronTC`.

---

### Phase 4: Solution Marching

Construct `GradientElement` per cell per substep with per-cell `tau`, and reconstruct `so1n` via the Phase 2.2 solve plus a weighting / limiter.

**Port from:**

- [Solver.hpp](https://github.com/solvcon/solvcon/blob/1e0eb87/libmarch/include/march/gas/Solver.hpp) -- `calc_so0t`, `calc_so0n`, `march`.
- [derivative.hpp](https://github.com/solvcon/solvcon/blob/1e0eb87/libmarch/include/march/gas/derivative.hpp) -- `calc_so1n`, `GradientWeigh`, `interpolate_solution`.
- [Jacobian.hpp](https://github.com/solvcon/solvcon/blob/1e0eb87/libmarch/include/march/gas/Jacobian.hpp) -- Euler flux Jacobian (2D `<4,2>`, 3D `<5,3>`).

**Files (actual):** `euler_march.cpp` (new), `euler_gradient.cpp` (new), `euler.hpp`, `cpp/modmesh/multidim/CMakeLists.txt`, `pymod/wrap_multidim.cpp`, `tests/test_multidim.py`.

**New parameters:** members `m_sigma0 = 3.0`, `m_taumin = 0.0`, `m_tauscale = 1.0` with read/write getters/setters.  Per cell: `acfl = |cflc(icl)|`, `sgm0 = sigma0/acfl`, `tau = taumin + acfl*tauscale`.

**Order-0 (`euler_march.cpp`):**

- `detail::EulerJacobian<NDIM>` (specialized for 2 and 3) -- `update(gamma, sol)` builds the flux Jacobian `jacos[ieq][jeq][idm]` and flux `fcn[ieq][idm]` from one conserved state (`TINY = 1e-60` guards the density divide); this is the inlined `Jacobian.hpp` port.
- `calc_solt()` -- `so0t = -(jacos . so1c)` per real cell.
- `calc_soln()` -- CESE space-time flux integral over the BCEs per real cell: spatial flux from the neighbor solution reconstructed at the BCE centroid times the BCE volume, minus the temporal flux summed over sub-faces (`qdt = dt/4`, `hdt = dt/2`), divided by the CCE volume `cevol(icl,0)`.  Neighbor cell via `msh.fcrcl(ifc, icl)`; the ghost-neighbor CE centroid falls back to `msh.clcnd(jcl, d)` when `jcl < 0` (matching `GradientElement`), so the `cecnd` ghost rows are not read.
- `calc_solt()` / `calc_soln()` dispatch on `m_ndim` (2 or 3, else throw) to the `detail::*_impl<NDIM>` templates.
- `march_substep()` -- order `update -> calc_solt -> calc_soln -> [bc_soln] -> calc_cfl -> calc_dsoln -> [bc_dsoln]`, with the two `bc_*` calls present as commented placeholders awaiting Phase 5.
- `march(steps)` -- runs `SUBSTEP_RUN = 2` substeps per step.  There is no simulation-time accumulator (solvcon's `time += dt/substep_run` is omitted); `time_increment` is the only time member.

**Order-1 (`euler_gradient.cpp`) -- `calc_dsoln()`:** per real cell, build `GradientElement(msh, m_cecnd, icl, tau)`, then for each fundamental gradient element (FGE): interpolate solution deltas `udf[ieq][ivx]` via a Taylor expansion about the neighbor (`so0c(jcl) + hdt*so0t(jcl) - so0n(icl)` plus `jdis(ifl,d) * so1c(jcl,ieq,d)`, with `jcl = gelem.rcl(ifl)`); solve `grad[ifge][ieq] = solve_gradient(ifge, udf[ieq])`; W-1/2 weighting (`alpha = 1`) `wgt = 1/sqrt(|grad|^2 + ALMOST_ZERO)` (`ALMOST_ZERO = 1e-200`) accumulated into `wacc[ieq]`; W-3/4 limiter `widv = wgt/wacc - ofg1` (`ofg1 = nfge_inverse()`), `sigma_max = min((1-ofg1)/wpa.max, -ofg1/wpa.min, sgm0)`; weighted reduction `so1n(icl,ieq,d) = sum_fge (ofg1 + sigma_max*widv) * grad`.  Fixed-size scratch uses `NFGE_MAX = GradientElementType::NFGE_MAX` and `NEQ_MAX = 5`.  The `interpolate_solution` step is inlined, not a named method.

**Bindings:** a new `wrap_march()` chained after `wrap_preparation()`, carrying `time_increment` (moved out of `wrap_management`), the `sigma0`/`taumin`/`tauscale` read/write properties, and `march` / `march_substep` / `update` / `calc_cfl` / `calc_solt` / `calc_soln` / `calc_dsoln` as `def_timed`.

**Tests (`_EulerMarchBase` across triangle/quad/mixed/tetra/hexa):** `test_march_param_defaults`; `test_calc_solt_euler_flux` (vs analytic flux via flux homogeneity); `test_calc_soln_freestream` (uniform state preserved); `test_calc_dsoln_linear_field` (exact gradient recovery); `test_calc_soln_against_reference` (independent finite-difference-Jacobian reimplementation -- constrains `hdt`/`qdt`, the temporal sign, and the Jacobian); `test_calc_dsoln_against_reference` (independent weighting/limiter reference, asserts the limiter is actually active); `test_march_bounded` (3 steps stay finite/positive, skipping single-cell meshes since without BCs an all-boundary mesh is not well posed).

---

### Phase 5: Boundary Conditions

Ghost-cell BCs applied as the `trim_do0`/`trim_do1` passes.  In libmarch "trim" means BC application, not a slope limiter or clamp.

**Port from:**

- [Trim.hpp](https://github.com/solvcon/solvcon/blob/1e0eb87/libmarch/include/march/gas/Trim.hpp) -- handler set NoOp / NonRefl / SlipWall / Inlet.
- [cese.py](https://github.com/solvcon/solvcon/blob/1e0eb87/solvcon/kerpak/cese.py) -- `bcsoln`/`bcdsoln` ordering.
- [gasdyn.py](https://github.com/solvcon/solvcon/blob/1e0eb87/solvcon/kerpak/gasdyn.py) -- `GasdynWall`/`GasdynInlet` (value names only).

**Handler math (from Trim.hpp).** Each handler writes the ghost row of a boundary face; `tfccls(ifc) = [interior_cell, ghost_cell]` and `mat = get_normal_matrix(ifc)` is the rotation with the outward face normal as its first axis.

- NoOp: both passes empty (default / symmetry handled elsewhere).
- NonRefl (also realizes outflow -- there is no separate Outlet handler): do0 copies `so0n(ghost) = so0n(interior)`; do1 rotates the interior order-1 solution into the normal frame, zeroes the normal-direction gradient (`vec[ieq][0] = 0`), keeps the tangential components, and rotates back (`so1n(ghost) = matinv . vec`).
- SlipWall: do0 copies density (col 0) and energy (col `neq-1`), then reflects momentum by rotating it into the normal frame, negating the normal component, and rotating back; do1 rotates the interior derivatives into the normal frame, negates the normal-direction derivatives and the normal-tangential cross terms, and rotates back.
- Inlet (supersonic, fixed Dirichlet) with value `[rho, v(ndim), p, gamma]`: do0 sets `so0n(ghost) = [rho, rho*v, p/(gamma-1) + 0.5*rho*|v|^2]` (the same primitive-to-conserved form as `init_solution`); do1 sets `so1n(ghost) = 0`.

**modmesh boundary mechanics (confirmed in `StaticMesh`).** Boundary faces live in `m_bndfcs` shape `[nbound, BFREL]`: col 0 = face index `ifc`, col 1 = boundary-group index, col 2 = related-block face (if any).  For boundary face `ifc`: interior cell `icl = fccls(ifc, 0)`; the ghost cell is the other side, reachable as `jcl = fccls(ifc, 1)` (negative) or `msh.fcrcl(ifc, icl)`.  Ghost cells are negative indices; ghost cell `igcl` maps back via `m_bndfcs(-igcl - 1, 0)`.  The outward unit normal is `m_fcnml(ifc, :)` (normalized; area is separate in `m_fcara`), and `fill_ghost` already relies on it pointing outward.  Named boundary groups are `StaticMeshBC` objects, each holding `facn() [nbound_group, BFREL]`.

**Files:** `euler_bc.cpp` (new), `euler.hpp`, `cpp/modmesh/multidim/CMakeLists.txt`, `pymod/wrap_multidim.cpp`, `tests/test_multidim.py`.

**Methods:** `get_normal_matrix(ifc)` (orthonormal frame with `fcnml(ifc,:)` as the first row; 2D `[[nx, ny], [-ny, nx]]`, 3D pick a stable tangent then complete with the cross product); `bc_soln()` (order-0 pass over boundary faces, inserted at the first commented hook in `march_substep`, after `calc_soln`); `bc_dsoln()` (order-1 pass, inserted at the second hook, after `calc_dsoln`).

**Tests:** NRBC ghost == interior (do0) and zero normal gradient (do1); slip wall reflects normal momentum and zeroes the wall-normal mass flux; inlet ghost == prescribed conserved state and zero ghost gradient; a walled 2D channel with inlet + non-reflective outflow stays finite/positive for several `march` steps (the first test where `march` runs with BCs, so it can replace the `ncell < 2` skip in `test_march_bounded`).

---

### Validation Milestone: Oblique Shock Reflection

After all phases: reproduce solvcon's oblique shock reflection on a 2D mesh (supersonic inlet + slip wall + NRBC).  The end-to-end validation is the scope of issue #404.

### PR DESCRIPTION
**Pull request by yungyuc**
This is the initial work for the project to port the multi-dimensional solver from solvcon to modmesh: https://github.com/orgs/solvcon/projects/9 .

Add `prepare_ce()` to compute conservation-element (CE) volumes, centroids, and sub-face metrics for 2D and 3D meshes.  Introduce `euler_ce.cpp` with template specializations `prepare_ce_2d<2>` and `prepare_ce_3d<3>`, expose the new arrays (`cevol`, `cecnd`, `sfcnd`, `sfnml`) through pybind11, add `StaticMesh::fcrcl` neighbor-cell helper, and add triangle-mesh unit tests.

The code was ported from https://github.com/solvcon/solvcon .  The old codebase contained multiple implementations.  We will verify the porting later.

For issue #66.

### CODE REVIEW

File: cpp/modmesh/mesh/StaticMesh.hpp [hunk 1]
```diff
@@ -327,6 +327,14 @@ class StaticMesh
      */
     int_type fcjcl(int_type ifc) const { return m_fccls(ifc, 1); }
 
+    /**
+     * Get the neighbor cell of @a icl across face @a ifc.
+     */
+    int_type fcrcl(int_type ifc, int_type icl) const
+    {
+        return m_fccls(ifc, 0) + m_fccls(ifc, 1) - icl;
```
Comment (yungyuc) [inline cpp/modmesh/mesh/StaticMesh.hpp:335]: One of the element is `icl`, so the expression returns the cell index of the other cell (neighbor).

File: cpp/modmesh/multidim/pymod/wrap_multidim.cpp [hunk 2]
```diff
@@ -64,7 +64,20 @@ class MODMESH_PYTHON_WRAPPER_VISIBILITY WrapEulerCore
                         return wrapped_type::construct(mesh, time_increment);
                     }),
                 py::arg("mesh"),
-                py::arg("time_increment"));
+                py::arg("time_increment"))
+            .def_property_readonly("ndim", &wrapped_type::ndim)
```
Comment (yungyuc) [inline cpp/modmesh/multidim/pymod/wrap_multidim.cpp:68]: Add code for prototyping.  Code will be organized in future PR.

File: cpp/modmesh/multidim/pymod/wrap_multidim.cpp [hunk 3]
```diff
@@ -64,7 +64,20 @@ class MODMESH_PYTHON_WRAPPER_VISIBILITY WrapEulerCore
...
+            .def_property_readonly("ncell", &wrapped_type::ncell)
+            .def_property_readonly("ngstcell", &wrapped_type::ngstcell)
+            .def_property_readonly("time_increment", &wrapped_type::time_increment)
+            .expose_SimpleArray("cevol", [](wrapped_type & self) -> decltype(auto)
```
Comment (yungyuc) [inline cpp/modmesh/multidim/pymod/wrap_multidim.cpp:72]: Expose `SimpleArray`.

File: cpp/modmesh/multidim/euler_ce.cpp [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+
+    m_sfnml = SimpleArray<real_type>(
+        std::vector<size_t>{
+            static_cast<size_t>(m_ncell),
+            StaticMesh::CLMFC * StaticMesh::FCMND,
+            static_cast<size_t>(m_ndim)},
+        0);
+}
+
+void EulerCore::prepare_ce()
```
Comment (yungyuc) [inline cpp/modmesh/multidim/euler_ce.cpp:82]: We will have only 2/3D for the Euler solver.

File: cpp/modmesh/multidim/euler_ce.cpp [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+    {
+        prepare_ce_3d();
+    }
+    else
+    {
+        throw std::invalid_argument("EulerCore: ndim must be 2 or 3");
+    }
+}
+
+void EulerCore::prepare_ce_2d()
```
Comment (yungyuc) [inline cpp/modmesh/multidim/euler_ce.cpp:98]: The old solvcon code used two different functions for 2/3D.  Keep this design for now.

File: tests/test_euler_multidim.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    def test_sfnml_shape(self):
+        self.assertEqual(
+            (self.ec.ncell,
+             modmesh.StaticMesh.CLMFC * modmesh.StaticMesh.FCMND,
+             self.ec.ndim),
+            self.ec.sfnml.shape)
+
+    def test_cce_volume_positive(self):
+        for icl in range(self.ec.ncell):
+            self.assertGreater(
```
Comment (yungyuc) [inline tests/test_euler_multidim.py:56]: Basic test.  When we validate the solver, this will be revisited.

File: tests/test_euler_multidim.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        assert_almost_equal(
+            self.ec.cevol[0, 0], 0.5 + 2.0 / 3.0 + 0.5, decimal=10)
+
+    def test_prepare_ce_rerun(self):
+        vol_before = self.ec.cevol[0, 0]
+        self.ec.prepare_ce()
+        assert_almost_equal(self.ec.cevol[0, 0], vol_before, decimal=12)
+
+
+@unittest.skip("3D ghost cell geometry has NaN due to fill_ghost bug")
```
Comment (yungyuc) [inline tests/test_euler_multidim.py:85]: Ghost cell will be ported in future PRs.
